editoast/editoast_models/src/model/error.rs

```diff
@@ -1,11 +1,58 @@
+use std::sync::LazyLock;
+
+use diesel::result::DatabaseErrorKind;
+use regex::Regex;
+
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    #[error("unique constraint violation: \"{constraint}\"")]
+    UniqueViolation { constraint: String },
+    #[error("check constraint violation on relation \"{relation}\": \"{constraint}\"")]
+    CheckViolation {
+        relation: String,
+        constraint: String,
+    },
     #[error(transparent)]
     DatabaseError(#[from] crate::DatabaseError),
 }
 
 impl From<diesel::result::Error> for Error {
     fn from(e: diesel::result::Error) -> Self {
-        Self::DatabaseError(e.into())
+        match &e {
```

Comment:
nit: not sure about my proposition: since all the workflow path ends up being `return                     Self::DatabaseError(e.into())` in case we can't do better, how about expressing this function with a bunch of early return possibilities, and always fallback to the default?

```rust
if let diesel::result::Error::DatabaseError(DatabaseErrorKind::UniqueViolation, inner) = &e {
  /* ... define RE ... */
  if let Some(captures) = RE.captures(something) {
    return Self::UniqueViolation { /* ... */ };
  } else {
    tracing::error!("...");
  }
}
/* same for CheckViolation */
return Self::DatabaseError(e.into());
```

Reply:
Oh, I was not entirely sure either. I just saw in the previous implementation, multiple possible outcomes of the function that were all doing the `Self::DatabaseError(e.into());`: they are all about "if I can't do anything, let's fallback to that".

My proposition is more about presenting the function with one `return` for each semantic outcome (`UniqueViolation`, `'CheckViolation` and fallback). But I could leave with any versions of the implementation, so feel free to keep or drop the commit.